File: src/frais/commands/update.py
from __future__ import annotations

import json
import logging
from typing import Annotated, Any

import typer
from rich.console import Console

from ..models import SourceKind, UpdateCandidate
from ..paths import ADVICE_CACHE
from ..plugins.base import ScannerPlugin

logger = logging.getLogger(__name__)
console = Console()


def _load_advice_cache_or_exit() -> Any:
    """Load the advice cache, or print error and raise Exit."""
    if not ADVICE_CACHE.exists():
        console.print("No advice cache found. Run [bold]frais advise[/bold] first.")
        raise typer.Exit(1)
    try:
        return json.loads(ADVICE_CACHE.read_text())
    except (json.JSONDecodeError, OSError) as exc:
        console.print(f"Failed to read advice cache: {exc}")
        raise typer.Exit(1)


def _parse_candidates_from_cache(data: Any) -> tuple[list[UpdateCandidate], dict[str, str]]:
    """Parse UpdateCandidate list and build id→plugin_name map from cache."""
    plugin_map: dict[str, str] = {}
    raw_candidates: list[dict[str, Any]] = []
    plugin_results: dict[str, Any] = data.get("plugin_results", {})
    for plugin_name, pr in plugin_results.items():
        for raw_cand in pr.get("candidates", []):
            item_data: dict[str, Any] = raw_cand.get("item", {})
            plugin_map[item_data.get("id", "")] = str(plugin_name)
        raw_candidates.extend(pr.get("candidates", []))

    candidates: list[UpdateCandidate] = []
    for raw in raw_candidates:
        try:
            candidates.append(UpdateCandidate.from_dict(raw))
        except Exception as exc:
            logger.warning("failed to parse cached candidate: %s", exc)
    return candidates, plugin_map


def _filter_candidates(candidates: list[UpdateCandidate], only: str | None) -> list[UpdateCandidate]:
    """Filter candidates by exact id or name match."""
    if not only:
        return candidates
    return [c for c in candidates if c.item.id == only or c.item.name == only]


def _execute_update_loop(
    candidates: list[UpdateCandidate],
    plugin_map: dict[str, str],
    plugins: dict[str, ScannerPlugin],
) -> None:
    """Interactive confirmation loop: display, confirm, execute."""
    for candidate in candidates:
        console.print()
        console.print(f"  [bold]{candidate.item.name}[/bold]  [dim]({candidate.item.id})[/dim]")
        console.print(
            f"  {candidate.item.current_version or 'unknown'} → "
            f"[green]{candidate.latest_version or 'unknown'}[/green]"
        )
        if candidate.ai_summary:
            from rich.markdown import Markdown
            console.print()
            console.print("  [bold cyan]AI Analysis[/]")
            console.print(Markdown(candidate.ai_summary))
        else:
            console.print()
            console.print(f"  [dim]No AI summary yet — `frais summarize {candidate.item.id}`[/dim]")

        if candidate.can_auto_update and candidate.item.source != SourceKind.APP_STORE:
            console.print(f"    [dim]cmd: {' '.join(candidate.command)}[/dim]")
        elif not candidate.can_auto_update:
            console.print("    [dim]manual update[/dim]")

        if not typer.confirm("  Proceed?", default=False):
            logger.info("update skipped name=%s", candidate.item.name)
            continue

        plugin_name = plugin_map.get(candidate.item.id)
        plugin = plugins.get(plugin_name) if plugin_name else None
        if plugin is not None:
            ok = plugin.update(candidate)
            logger.info("update executed plugin=%s name=%s ok=%s", plugin_name, candidate.item.name, ok)


def update(
    only: Annotated[
        str | None,
        typer.Argument(
            help="Filter by exact id or software name. Omit to review all candidates.",
            metavar="ID_OR_NAME",
        ),
    ] = None,
) -> None:
    """Interactively review and execute updates with AI advice.

    Loads results from the last `frais advise` run. Shows each candidate
    with AI advice for confirmation. Auto-updatable packages (Homebrew, npm)
    execute directly; others show the recommended action.

    Run `frais advise` first to generate the update candidates.

    Examples:
      frais update
      frais update npm
    """
    from ..plugins.registry import all_plugins

    data = _load_advice_cache_or_exit()
    candidates, plugin_map = _parse_candidates_from_cache(data)

    candidates = _filter_candidates(candidates, only)
    if not candidates:
        console.print("No update candidates found.")
        return

    plugins = all_plugins()
    _execute_update_loop(candidates, plugin_map, plugins)

File: src/cli/update.rs
// update command — matches Python's frais/commands/update.py.
use super::UpdateArgs;
use crate::models::UpdateCandidate;
use std::collections::BTreeMap;

pub fn run(args: UpdateArgs) -> Result<(), String> {
    // Load advice cache — matches Python's _load_advice_cache_or_exit()
    let cache_path = crate::paths::advice_cache();
    if !cache_path.exists() {
        println!("No advice cache found. Run `frais advise` first.");
        return Err("No advice cache found.".into());
    }

    let data: serde_json::Value = match std::fs::read_to_string(&cache_path) {
        Ok(content) => match serde_json::from_str(&content) {
            Ok(v) => v,
            Err(e) => {
                println!("Failed to read advice cache: {e}");
                return Err(format!("Failed to read advice cache: {e}"));
            }
        },
        Err(e) => {
            println!("Failed to read advice cache: {e}");
            return Err(format!("Failed to read advice cache: {e}"));
        }
    };

    // Parse candidates and build id→plugin_name map — matches _parse_candidates_from_cache()
    let (candidates, plugin_map) = parse_candidates_from_cache(&data);

    // Filter by exact id or name — matches _filter_candidates()
    let candidates = filter_candidates(&candidates, args.only.as_deref());
    if candidates.is_empty() {
        println!("No update candidates found.");
        return Ok(());
    }

    // Execute interactive update loop
    let plugins = crate::plugins::registry::all_plugins();
    execute_update_loop(&candidates, &plugin_map, &plugins);

    Ok(())
}

/// Parse UpdateCandidate list and build id→plugin_name map from cache.
/// Matches Python's _parse_candidates_from_cache().
fn parse_candidates_from_cache(
    data: &serde_json::Value,
) -> (Vec<UpdateCandidate>, BTreeMap<String, String>) {
    let mut plugin_map: BTreeMap<String, String> = BTreeMap::new();
    let mut candidates: Vec<UpdateCandidate> = Vec::new();

    let plugin_results = match data.get("plugin_results").and_then(|v| v.as_object()) {
        Some(pr) => pr,
        None => return (candidates, plugin_map),
    };

    for (plugin_name, pr) in plugin_results {
        let raw_candidates = match pr.get("candidates").and_then(|v| v.as_array()) {
            Some(arr) => arr,
            None => continue,
        };

        for raw_cand in raw_candidates {
            // Build plugin_map from item.id in each candidate
            if let Some(item) = raw_cand.get("item") {
                if let Some(item_id) = item.get("id").and_then(|v| v.as_str()) {
                    plugin_map.insert(item_id.to_string(), plugin_name.clone());
                }
            }

            // Parse UpdateCandidate from dict
            match serde_json::from_value::<UpdateCandidate>(raw_cand.clone()) {
                Ok(cand) => candidates.push(cand),
                Err(e) => {
                    log::warn!("failed to parse cached candidate: {}", e);
                }
            }
        }
    }

    (candidates, plugin_map)
}

/// Filter candidates by exact id or name match.
/// Matches Python's _filter_candidates() — EXACT match, not substring.
fn filter_candidates(
    candidates: &[UpdateCandidate],
    only: Option<&str>,
) -> Vec<UpdateCandidate> {
    match only {
        None => candidates.to_vec(),
        Some(filter) => candidates
            .iter()
            .filter(|c| c.item.id == filter || c.item.name == filter)
            .cloned()
            .collect(),
    }
}

/// Interactive confirmation loop: display, confirm, execute.
/// Matches Python's _execute_update_loop().
fn execute_update_loop(
    candidates: &[UpdateCandidate],
    plugin_map: &BTreeMap<String, String>,
    plugins: &BTreeMap<String, Box<dyn crate::plugins::ScannerPlugin>>,
) {
    for candidate in candidates {
        println!();
        println!(
            "  {}  ({})",
            candidate.item.name, candidate.item.id
        );
        println!(
            "  {} → {}",
            candidate.item.current_version.as_deref().unwrap_or("unknown"),
            candidate.latest_version.as_deref().unwrap_or("unknown"),
        );
        if let Some(ref summary) = candidate.ai_summary {
            println!();
            println!("  AI Analysis");
            // Render markdown in terminal — matches Python's console.print(Markdown(candidate.ai_summary))
            let skin = termimad::MadSkin::default();
            // Indent each line by 2 spaces to match Python's output
            let indented = summary
                .lines()
                .map(|l| format!("  {}", l))
                .collect::<Vec<_>>()
                .join("\n");
            skin.print_text(&indented);
        } else {
            println!();
            println!(
                "  No AI summary yet — `frais summarize {}`",
                candidate.item.id
            );
        }

        if candidate.can_auto_update
            && candidate.item.source != crate::models::SourceKind::AppStore
        {
            println!("    cmd: {}", candidate.command.join(" "));
        } else if !candidate.can_auto_update {
            println!("    manual update");
        }

        // Confirm — matches typer.confirm("  Proceed?", default=False)
        print!("  Proceed? [y/N]: ");
        use std::io::Write;
        let _ = std::io::stdout().flush();
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).ok();
        if input.trim().to_lowercase() != "y" {
            log::info!("update skipped name={}", candidate.item.name);
            continue;
        }

        // Look up plugin from cache-built map — matches Python's plugin_map.get(candidate.item.id)
        let plugin_name = plugin_map.get(&candidate.item.id);
        let plugin = plugin_name.and_then(|n| plugins.get(n));
        if let Some(plugin) = plugin {
            let ok = plugin.update(candidate);
            log::info!(
                "update executed plugin={} name={} ok={}",
                plugin_name.unwrap_or(&String::new()),
                candidate.item.name,
                ok
            );
        } else {
            log::warn!(
                "update skipped: no plugin found for candidate {} (plugin_name={:?}, plugin_map keys={:?})",
                candidate.item.id,
                plugin_name,
                plugin_map.keys().collect::<Vec<_>>()
            );
            eprintln!("  Warning: no plugin found for {} — the cache may be corrupt. Try running `frais advise` again.", candidate.item.id);
        }
    }

    println!();
}
